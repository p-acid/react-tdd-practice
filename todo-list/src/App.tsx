import { AppBox, Container, Form, TodoListWrapper } from 'AppStyle';
import { Button, Input, TodoItem } from 'components';
import { FormEvent, useCallback, useRef, useState } from 'react';

function App() {
  const [todoList, setTodoList] = useState<string[]>([]);
  const inputRef = useRef<HTMLInputElement>(null);

  const addTodo = useCallback(
    (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      const inputValue = inputRef.current?.value;

      if (!inputValue) return;

      setTodoList((prev) => {
        if (prev.includes(inputValue)) return prev;

        return [...prev, inputValue];
      });

      event.currentTarget.reset();
    },
    [setTodoList, inputRef],
  );

  const removeTodo = useCallback(
    (content: string) => {
      setTodoList((prev) => prev.filter((todoItem) => todoItem !== content));
    },
    [setTodoList],
  );

  return (
    <Container>
      <AppBox>
        <TodoListWrapper data-testid="todoList">
          {todoList.map((content) => (
            <TodoItem key={`todoItem-${content}`} onRemove={() => removeTodo(content)}>
              {content}
            </TodoItem>
          ))}
        </TodoListWrapper>
        <Form onSubmit={addTodo}>
          <Input ref={inputRef} placeholder="할 일을 입력하세요" />
          <Button>추가</Button>
        </Form>
      </AppBox>
    </Container>
  );
}

export default App;
