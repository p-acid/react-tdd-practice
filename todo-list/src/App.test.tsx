import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';

import App from 'App';
import { Button, Input } from 'components';

describe('<App />', () => {
  it('renders component correctly', () => {
    render(<App />);

    expect(screen).toMatchSnapshot();
  });
  it('renders Button component correctly', () => {
    const handleClick = jest.fn();

    render(<Button onClick={handleClick}>추가</Button>);

    const buttonElement = screen.getByText('추가');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyleRule('background-color', '#111111');
    expect(buttonElement).toHaveStyleRule('background-color', '#1a1a1a', {
      modifier: ':hover',
    });

    expect(handleClick).toHaveBeenCalledTimes(0);
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);

    expect(buttonElement).toMatchSnapshot();
  });
  it('renders Input component correctly', () => {
    const handleChange = jest.fn();

    render(
      <Input value="default value" placeholder="default placeholder" onChange={handleChange} />,
    );

    const inputElement = screen.getByDisplayValue('default value');

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute('placeholder', 'default placeholder');

    expect(inputElement).toMatchSnapshot();
  });
  it('change Input component value', () => {
    const handleChange = jest.fn();

    render(<Input placeholder="default placeholder" onChange={handleChange} />);

    const inputElement: HTMLInputElement = screen.getByPlaceholderText('default placeholder');

    fireEvent.change(inputElement, { target: { value: 'changed value' } });
    expect(inputElement.value).toBe('changed value');

    expect(inputElement).toBeInTheDocument();
  });
});
