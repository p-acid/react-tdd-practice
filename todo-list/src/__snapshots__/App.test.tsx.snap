// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders Button component correctly 1`] = `
.c0 {
  padding: 12px;
  border: none;
  border-radius: 8px;
  color: #dfdfdf;
  background-color: #111111;
  cursor: pointer;
  word-break: keep-all;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0:hover {
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  background-color: #1a1a1a;
}

.c0:active {
  -webkit-transform: translateY(0px);
  -ms-transform: translateY(0px);
  transform: translateY(0px);
}

<button
  class="c0"
>
  추가
</button>
`;

exports[`<App /> renders Input component correctly 1`] = `
.c0 {
  padding: 12px;
  border: 1px solid #bdbdbd;
  border-radius: 8px;
  color: #131313;
  background-color: #ffffff;
  outline: none;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0:focus {
  box-shadow: 1px black;
}

<input
  class="c0"
  placeholder="default placeholder"
  value="default value"
/>
`;

exports[`<App /> renders component correctly 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`<TodoItem /> renders component correctly 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
